pkg/healthcheck/addons.go

```diff
@@ -0,0 +1,157 @@
+package healthcheck
+
+import (
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/ghodss/yaml"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/retry"
+
+	acsapi "github.com/openshift/openshift-azure/pkg/api"
+)
+
+// TODO: consider if this should be a separate function in the plugin interface
+func ensureSyncPod(kc *kubernetes.Clientset, cs *acsapi.OpenShiftManagedCluster) error {
+	csb, err := yaml.Marshal(cs)
+	if err != nil {
+		return err
+	}
+
+	kcb, err := yaml.Marshal(cs.Config.SyncKubeconfig)
+	if err != nil {
+		return err
+	}
+
+	// TODO: Hash content and add as an annotation in the deployment template
+	syncSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "sync",
+			Namespace: "openshift-infra",
+		},
+		Data: map[string][]byte{
+			"containerservice.yaml": csb,
+			"sync.kubeconfig":       kcb,
+		},
+	}
+
+	_, err = kc.CoreV1().Secrets(syncSecret.Namespace).Create(syncSecret)
+	switch {
+	case err == nil:
+
+	case errors.IsAlreadyExists(err):
+		err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			existingSecret, err := kc.CoreV1().Secrets(syncSecret.Namespace).Get(syncSecret.Name, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+
+			syncSecret.ResourceVersion = existingSecret.ResourceVersion
+			_, err = kc.CoreV1().Secrets(syncSecret.Namespace).Update(syncSecret)
+			return err
+		})
+		if err != nil {
+			return err
+		}
+
+	default:
+		return err
+	}
+
+	syncDeployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "sync",
+			Namespace: "openshift-infra",
```

Comment:
actually it turns out that the sync pod will have to go into its own dedicated namespace which is ignored by the sync pod - otherwise the RP could update the sync pod and be countermanded by the running sync pod before it's killed...